From c8a4a9a438f79e8cf1d3479f91098c13e10c4268 Mon Sep 17 00:00:00 2001
From: jinghua <jinghua@marvell.com>
Date: Mon, 30 Nov 2015 14:19:20 +0800
Subject: [PATCH 0545/1240] fix: spi: a3700: fix missing
 north_bridge_test_pin_output_en for SPI in SPL

    for SPI to work in UART boot mode, there is a test_pin which has to be set,
    that pin switch two SPI pins back to be used by SPI as clk and CS.
    These two pins are used by uart in uart boot mode, to get singles as input.
    It is not in any document yet. information referenced from designer.

Change-Id: Ie626ad5abcc4c3bffc106f616b7bb2149bff7615
Signed-off-by: jinghua <jinghua@marvell.com>
Change-Id: Idac6981a6114265e5e4807761fda1f9fe0088eb3
Reviewed-on: http://vgitil04.il.marvell.com:8080/25396
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
---
 arch/arm/cpu/armv8/armadalp/mvebu_gpio.c       | 12 ++++++++++++
 arch/arm/include/asm/arch-armadalp/regs-base.h |  3 +++
 2 files changed, 15 insertions(+)

diff --git a/arch/arm/cpu/armv8/armadalp/mvebu_gpio.c b/arch/arm/cpu/armv8/armadalp/mvebu_gpio.c
index 182e9d4..e48be96 100644
--- a/arch/arm/cpu/armv8/armadalp/mvebu_gpio.c
+++ b/arch/arm/cpu/armv8/armadalp/mvebu_gpio.c
@@ -56,6 +56,18 @@ void mvebu_a3700_gpio(void)
 	reg_val = reg_val & (~(1 << MVEBU_A3700_GPIO_SPI_GPIO_EN_OFF));
 	writel(reg_val, MVEBU_A3700_GPIO_NB_SEL);
 
+	/* set hiden GPIO setting for SPI
+	  * in north_bridge_test_pin_out_en register 13804,
+	  * bit 28 is the one which enables CS, CLK pin to be
+	  * output, need to set it to 1.
+	  * it is not in any document, but in UART boot mode,
+	  * CS, CLK pin will be twisted and be used for input.
+	  * which breaks SPI functionality.
+	  */
+	reg_val = readl(MVEBU_A3700_NB_TEST_PIN_OUTPUT_EN);
+	reg_val = reg_val | (1 << MVEBU_A3700_NB_TEST_PIN_OUTPUT_SPI_EN_OFF);
+	writel(reg_val, MVEBU_A3700_NB_TEST_PIN_OUTPUT_EN);
+
 	return;
 }
 
diff --git a/arch/arm/include/asm/arch-armadalp/regs-base.h b/arch/arm/include/asm/arch-armadalp/regs-base.h
index 6050037..1ad0a45 100644
--- a/arch/arm/include/asm/arch-armadalp/regs-base.h
+++ b/arch/arm/include/asm/arch-armadalp/regs-base.h
@@ -51,6 +51,9 @@
 #define MVEBU_A3700_GPIO_TW1_GPIO_EN_OFF (10)
 #define MVEBU_A3700_GPIO_SPI_GPIO_EN_OFF (18)
 
+#define MVEBU_A3700_NB_TEST_PIN_OUTPUT_EN (MVEBU_REGS_BASE + 0x13804)
+#define MVEBU_A3700_NB_TEST_PIN_OUTPUT_SPI_EN_OFF (28)
+
 /* clock registers */
 #define MVEBU_NORTH_CLOCK_REGS_BASE	(MVEBU_REGS_BASE + 0x13000)
 #define MVEBU_SOUTH_CLOCK_REGS_BASE	(MVEBU_REGS_BASE + 0x18000)
-- 
1.9.1

